### PR TITLE
refactor: Refactor data scratch-Android run

### LINKED ISSUE #1754: Data scratch - android test matrix
# Plan

According to the specification described in epic [#1727](https://github.com/Flank/flank/issues/1727)
add abstraction and implementation for the following part of the data layer:

## Abstraction

`ftl/data/TestMatrixAndroid.kt`

```kotlin
package ftl.data

val executeTestMatrixAndroid: TestMatrixAndroid.Execute get() = TODO()

object TestMatrixAndroid {

    data class Config(
        // args
        val clientDetails: Map<String, String>?,
        val resultsBucket: String,
        val autoGoogleLogin: Boolean,
        val networkProfile: String?,
        val directoriesToPull: List<String>,
        val obbNames: List<String>,
        val environmentVariables: Map<String, String>,
        val autograntPermissions: Boolean,
        val testTimeout: String,
        val performanceMetrics: Boolean,
        val recordVideo: Boolean,
        val flakyTestAttempts: Int,
        val failFast: Boolean,
        val project: String,
        val resultsHistoryName: String?,

        // build
        val otherFiles: Map<String, String>,
        val runGcsPath: String,
        val devices: List<Device>,
        val additionalApkGcsPaths: List<String>,
        val obbFiles: Map<String, String>,
    )

    sealed class Type {
        data class Instrumentation(
            val appApkGcsPath: String,
            val testApkGcsPath: String,
            val testRunnerClass: String?,
            val orchestratorOption: String?,
            // sharding
            val disableSharding: Boolean,
            val testShards: ShardChunks,
            val numUniformShards: Int?,
            val keepTestTargetsEmpty: Boolean,
            val environmentVariables: Map<String, String> = emptyMap(),
            val testTargetsForShard: ShardChunks
        ) : Type()

        data class Robo(
            val appApkGcsPath: String,
            val flankRoboDirectives: List<FlankRoboDirective>?,
            val roboScriptGcsPath: String?
        ) : Type()

        data class GameLoop(
            val appApkGcsPath: String,
            val testRunnerClass: String?,
            val scenarioNumbers: List<String>,
            val scenarioLabels: List<String>
        ) : Type()
    }

    interface Execute : (Config, Type) -> TestMatrix.Result
}
```

## Target

* `AndroidRunCommand` -> `AndroidArgs/runAndroidTests` -> `GcAndroidTestMatrix/build`
* `AndroidTestConfig`

## Adapter

`ftl/adapter/GoogleTestMatrixAndroid.kt`

### PR DESCRIPTION
**Pull request by piotradamczyk5**
Fixes #1754

- code is refactored according to the description provided in #1754 with a change 
    `interface Execute : (Config, Type) -> TestMatrix.Result` ->  `interface Execute : (Config, List<Type>) -> <TestMatrix.Data>`
- all runs of `flank firebase test android run ... ` work without any problems like previously

## Checklist

- [x] Unit tested

### CODE REVIEW

**Comment by piotradamczyk5** [comment]
@flank-it

**Comment by piotradamczyk5** [comment]
@flank-it

File: test_runner/src/main/kotlin/ftl/run/platform/RunAndroidTests.kt [hunk 1]
```diff
@@ -23,82 +17,55 @@ import ftl.run.model.TestResult
...
-    if (testMatrices.isEmpty()) throw FlankGeneralError("There are no tests to run.")
+        .map { context -> createAndroidTestMatrixType(context) }
+        .run { executeTestMatrixAndroid(createAndroidTestConfig(args), toList()) }
+        .takeIf { it.isNotEmpty() }
+        ?: throw FlankGeneralError("There are no tests to run.")
 
     logLn(beforeRunMessage(allTestShardChunks))
 
     TestResult(
         // TDODO: https://github.com/Flank/flank/issues/1754
```
Comment (pawelpasterz) [inline test_runner/src/main/kotlin/ftl/run/platform/RunAndroidTests.kt:56]: To be removed?

**Review by pawelpasterz — APPROVED** [review]

File: test_runner/src/main/kotlin/ftl/run/platform/RunAndroidTests.kt (hunk 1 above)
Comment (piotradamczyk5) [reply test_runner/src/main/kotlin/ftl/run/platform/RunAndroidTests.kt:56]:
```suggestion
```

**Review by jan-gogo — COMMENTED** [review]
Looks really nice 👍. One thing, the stuff from `ftl.gc.android` is operating on `com.google.testing` lib so it should be in `ftl.client.google` package or nested, but definitely in the client layer.

**Comment by piotradamczyk5** [comment]
> Looks really nice 👍. One thing, the stuff from `ftl.gc.android` is operating on `com.google.testing` lib so it should be in `ftl.client.google` package or nested, but definitely in the client layer.

Nice catch, move to nested package because there is a little mess in `ftl.client.google`

**Review by pawelpasterz — APPROVED** [review]

**Review by pawelpasterz — APPROVED** [review]

**Review by Sloox — APPROVED** [review]

**Review by Sloox — APPROVED** [review]

**Review by Sloox — APPROVED** [review]

**Pull request merged**